core/pkg/artifacts/saver.go

```
@@ -295,10 +329,164 @@ func newUploadTask(fileInfo serverFileResponse, localPath string) *filetransfer.
+		return nil, nil
+	}
+	if fileSize > S3MaxMultiUploadSize {
+		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	partsInfo := []gql.UploadPartsInput{}
+	partNumber := int64(1)
+	buffer := make([]byte, getChunkSize(fileSize))
+	for {
+		bytesRead, err := file.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if bytesRead == 0 {
+			break
+		}
+		partsInfo = append(partsInfo, gql.UploadPartsInput{
+			PartNumber: partNumber,
+			HexMD5:     utils.ComputeHexMD5(buffer[:bytesRead]),
+		})
+		partNumber++
+	}
+	return partsInfo, nil
+}
+
+func (as *ArtifactSaver) uploadMultipart(
+	path string,
+	fileInfo serverFileResponse,
+	partData []gql.UploadPartsInput,
+	doneChan chan<- uploadResult,
+) {
+	statInfo, err := os.Stat(path)
+	if err != nil {
+		doneChan <- uploadResult{name: fileInfo.name, err: err}
+		return
+	}
+	chunkSize := getChunkSize(statInfo.Size())
+
+	type partResponse struct {
+		partNumber int64
+		task       *filetransfer.Task
+	}
+
+	wg := sync.WaitGroup{}
+	subChan := make(chan partResponse, len(partData))
+	// TODO: add mid-upload cancel.
+
+	var contentType string
+	for _, h := range fileInfo.uploadHeaders {
+		if strings.HasPrefix(h, "Content-Type:") {
+			contentType = strings.TrimPrefix(h, "Content-Type:")
+			break
+		}
+	}
+	if contentType == "" {
+		err := fmt.Errorf("content-type header is required for multipart uploads")
+		doneChan <- uploadResult{name: fileInfo.name, err: err}
+		return
+	}
+
+	partInfo := fileInfo.multipartUploadInfo
+	for i, part := range partInfo {
+		task := newUploadTask(fileInfo, path)
+		task.Url = part.UploadUrl
+		task.Offset = int64(i) * chunkSize
+		remainingSize := statInfo.Size() - task.Offset
+		task.Size = min(remainingSize, chunkSize)
+		b64md5, err := utils.HexToB64(partData[i].HexMD5)
+		if err != nil {
+			panic(err)
+		}
```

important: plain panics are not allowed. In this case, it looks like you can return an error instead.

---

Sorry, this must have made it past some debugging I was doing. Fixed.